qiskit/circuit/quantumcircuit.py

```diff
@@ -757,16 +762,25 @@ def qasm(self, formatted=False, filename=None):
             # If instruction is a composite circuit
             elif not isinstance(instruction, Gate) and (instruction.name not in ['barrier',
                                                                                  'reset']):
-                qasm_string = self._get_composite_circuit_qasm_from_instruction(instruction)
+                if instruction not in existing_composite_circuits:
+                    if instruction.name in existing_gate_names:
+                        instruction.name += "_" + str(id(self))
+
+                        warnings.warn("A gate named {} already exists. "
+                                      "We have renamed "
+                                      "your gate to {}".format(instruction.name.split("_")[0],
```

Comment:
This is hacky way to get the "old" name and error prune. For example `my_gate` raises a message `A gate named my already exists. We have renamed your gate to my_gate_47432`. Notice `... named my already ...`.

I think it is just simpler to do this:
```suggestion
                        old_name = instruction.name
                        instruction.name += "_" + str(id(self))

                        warnings.warn("A gate named {} already exists. "
                                      "We have renamed "
                                      "your gate to {}".format(old_name,
```

Reply:
For some reason, id is not unique here (dunno why):
```
my_gate = QuantumCircuit(1, name='my_gate')
my_gate.h(0)
my_gate_inst1 = my_gate.to_instruction()

my_gate = QuantumCircuit(1, name='my_gate')
my_gate.x(0)
my_gate_inst2 = my_gate.to_instruction()

my_gate = QuantumCircuit(1, name='my_gate')
my_gate.x(0)
my_gate_inst3 = my_gate.to_instruction()

qr = QuantumRegister(1, name='qr')
circuit = QuantumCircuit(qr, name='circuit')
circuit.append(my_gate_inst1, [qr[0]])
circuit.append(my_gate_inst2, [qr[0]])
circuit.append(my_gate_inst3, [qr[0]])

print(circuit.qasm())
```
OPENQASM 2.0;
include "qelib1.inc";
gate my_gate_4737533648 q0 {x q0; }
gate my_gate_4737533648 q0 {x q0; }
gate my_gate q0 {h q0; }
qreg qr[1];
my_gate qr[0];
my_gate_4737533648 qr[0];
my_gate_4737533648 qr[0];
```
Maybe Im missing something :-/

Reply:
I added the change you suggested with the `instruction.name`. The non-unique id's was my mistake. I had set it up that it was attaching an `id` of the circuit instead of the instruction. I have fixed this and the id's should now be unique.